crates/runtime/src/backtrace.rs

```diff
@@ -0,0 +1,153 @@
+//! Backtrace object and utilities.
+
+use crate::jit_function_registry;
+use std::sync::Arc;
+
+/// Information about backtrace frame.
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+pub struct BacktraceFrame {
+    pc: usize,
+}
+
+impl Default for BacktraceFrame {
+    fn default() -> Self {
+        Self { pc: 0 }
+    }
+}
+
+impl BacktraceFrame {
+    /// Current PC or IP pointer for the frame.
+    pub fn pc(&self) -> usize {
+        self.pc
+    }
+    /// Additinal frame information.
+    pub fn tag(&self) -> Option<Arc<jit_function_registry::JITFunctionTag>> {
+        jit_function_registry::find(self.pc)
+    }
+}
+
+const BACKTRACE_LIMIT: usize = 32;
+
+/// Backtrace during WebAssembly trap.
+#[derive(Copy, Clone, PartialEq, Eq)]
+pub struct Backtrace {
+    len: usize,
+    frames: [BacktraceFrame; BACKTRACE_LIMIT],
+}
+
+impl Backtrace {
+    fn new() -> Self {
+        Self {
+            len: 0,
+            frames: [Default::default(); BACKTRACE_LIMIT],
+        }
+    }
+    fn full(&self) -> bool {
+        self.len >= BACKTRACE_LIMIT
+    }
+    fn push(&mut self, frame: BacktraceFrame) {
+        assert!(self.len < BACKTRACE_LIMIT);
+        self.frames[self.len] = frame;
+        self.len += 1;
+    }
+    /// Amount of the backtrace frames.
+    pub fn len(&self) -> usize {
+        self.len
+    }
+}
+
+impl std::ops::Index<usize> for Backtrace {
+    type Output = BacktraceFrame;
+    fn index(&self, index: usize) -> &Self::Output {
+        assert!(index < self.len);
+        &self.frames[index]
+    }
+}
+
+impl std::fmt::Debug for Backtrace {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        writeln!(f, "Backtrace![")?;
+        for i in 0..self.len() {
+            let frame = &self.frames[i];
+            writeln!(f, "  {:x}: {:?}", frame.pc(), frame.tag())?;
+        }
+        write!(f, "]")?;
+        Ok(())
+    }
+}
+
+#[cfg(not(all(target_os = "windows", target_arch = "x86_64")))]
+fn capture_stack<F>(mut f: F)
+where
+    F: FnMut(usize) -> bool,
+{
+    use backtrace::trace;
+    trace(|frame| {
+        let pc = frame.ip() as usize;
+        f(pc)
+    });
+}
+
+#[cfg(all(target_os = "windows", target_arch = "x86_64"))]
+fn capture_stack<F>(mut f: F)
+where
+    F: FnMut(usize) -> bool,
+{
+    use std::mem::MaybeUninit;
+    use std::ptr;
+    use winapi::um::winnt::{
+        RtlCaptureContext, RtlLookupFunctionEntry, RtlVirtualUnwind, CONTEXT, UNWIND_HISTORY_TABLE,
+        UNW_FLAG_NHANDLER,
+    };
+
+    #[repr(C, align(16))]
+    struct WrappedContext(CONTEXT);
+    #[repr(C, align(16))]
+    struct WrappedUnwindHistory(UNWIND_HISTORY_TABLE);
```

Comment:
`UNWIND_HISTORY_TABLE` does not appear to be `__declspec(align(16))` like `CONTEXT` is, so I don't think this wrapper is necessary and we can use the `winapi` defined type without issue.